[Navigation] Throw exceptions when unloading

rwlbuis opened the pull request:
#### 3209febcce786a323163ed4c5a67c96057507e1e
<pre>
[Navigation] Throw exceptions when unloading
<a href="https://bugs.webkit.org/show_bug.cgi?id=276176">https://bugs.webkit.org/show_bug.cgi?id=276176</a>

Reviewed by Alex Christensen.

Throw exceptions when unloading as specified for multiple methods [1, 2, 3].

This PR also renames m_ignoreOpensDuringUnloadCount to better reflect the specification [4]

[1] <a href="https://html.spec.whatwg.org/multipage/nav-history-apis.html#dom-navigation-navigate">https://html.spec.whatwg.org/multipage/nav-history-apis.html#dom-navigation-navigate</a>
[2] <a href="https://html.spec.whatwg.org/multipage/nav-history-apis.html#dom-navigation-reload">https://html.spec.whatwg.org/multipage/nav-history-apis.html#dom-navigation-reload</a>
[3] <a href="https://html.spec.whatwg.org/multipage/nav-history-apis.html#performing-a-navigation-api-traversal">https://html.spec.whatwg.org/multipage/nav-history-apis.html#performing-a-navigation-api-traversal</a>
[4] <a href="https://html.spec.whatwg.org/multipage/document-lifecycle.html#unload-counter">https://html.spec.whatwg.org/multipage/document-lifecycle.html#unload-counter</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/navigate-pagehide-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/reload-pagehide-expected.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::open):
(WebCore::Document::write):
* Source/WebCore/dom/Document.h:
(WebCore::Document::unloadCounter const):
* Source/WebCore/dom/UnloadCountIncrementer.h: Renamed from Source/WebCore/dom/IgnoreOpensDuringUnloadCountIncrementer.h.
(WebCore::UnloadCountIncrementer::UnloadCountIncrementer):
(WebCore::UnloadCountIncrementer::~UnloadCountIncrementer):
* Source/WebCore/history/BackForwardCache.cpp:
(WebCore::firePageHideEventRecursively):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::detachChildren):
(WebCore::FrameLoader::shouldClose):
(WebCore::FrameLoader::dispatchUnloadEvents):
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::reload):
(WebCore::Navigation::navigate):
(WebCore::Navigation::performTraversal):

Canonical link: <a href="https://commits.webkit.org/280898@main">https://commits.webkit.org/280898@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8fa65a4ca83a8ad599f415d0d3cf183608c86390

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57759 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37087 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10235 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61381 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8204 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59887 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44723 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8392 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46804 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5822 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59789 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34786 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49914 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27630 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31556 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7207 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7208 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53508 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7477 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63064 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1673 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7551 "Found 1 new test failure: imported/w3c/web-platform-tests/html/semantics/embedded-content/media-elements/loading-the-media-resource/resource-selection-invoke-pause-networkState.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54020 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1679 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49925 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54138 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12823 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1420 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32916 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34002 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35086 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33747 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/dom/IgnoreOpensDuringUnloadCountIncrementer.h:
```diff
@@ -33,7 +33,7 @@ class IgnoreOpensDuringUnloadCountIncrementer {
     WTF_MAKE_NONCOPYABLE(IgnoreOpensDuringUnloadCountIncrementer);
 public:
     explicit IgnoreOpensDuringUnloadCountIncrementer(Document* document)
```
This should probably be at least renamed.  I actually think we should replace its unsigned* with a ScopeExit that captures a WeakPtr&lt;Document&gt;

rwlbuis replied:
I  renamed the helper class. I am not really seeing how to apply the ScopeExit here, but anyway I think it would remove the nice ctor/dtor symmetry. I do prefer a bit to hold the Document rather than an unsigned*, so I made that change.

rwlbuis committed:
[Navigation] Throw exceptions when unloading
https://bugs.webkit.org/show_bug.cgi?id=276176

Reviewed by Alex Christensen.

Throw exceptions when unloading as specified for multiple methods [1, 2, 3].

This PR also renames m_ignoreOpensDuringUnloadCount to better reflect the specification [4]

[1] https://html.spec.whatwg.org/multipage/nav-history-apis.html#dom-navigation-navigate
[2] https://html.spec.whatwg.org/multipage/nav-history-apis.html#dom-navigation-reload
[3] https://html.spec.whatwg.org/multipage/nav-history-apis.html#performing-a-navigation-api-traversal
[4] https://html.spec.whatwg.org/multipage/document-lifecycle.html#unload-counter

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/navigate-pagehide-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/reload-pagehide-expected.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::open):
(WebCore::Document::write):
* Source/WebCore/dom/Document.h:
(WebCore::Document::unloadCounter const):
* Source/WebCore/dom/UnloadCountIncrementer.h: Renamed from Source/WebCore/dom/IgnoreOpensDuringUnloadCountIncrementer.h.
(WebCore::UnloadCountIncrementer::UnloadCountIncrementer):
(WebCore::UnloadCountIncrementer::~UnloadCountIncrementer):
* Source/WebCore/history/BackForwardCache.cpp:
(WebCore::firePageHideEventRecursively):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::detachChildren):
(WebCore::FrameLoader::shouldClose):
(WebCore::FrameLoader::dispatchUnloadEvents):
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::reload):
(WebCore::Navigation::navigate):
(WebCore::Navigation::performTraversal):

Canonical link: https://commits.webkit.org/280898@main
